docs/performance/inserts/architecture.txt

```diff
@@ -0,0 +1,263 @@
+expressions, as listed under the previous header.
+
+Additionally, the planner prepares a bulk insert response object that will hold
+the results of every row we try to insert.
+
+.. _execution:
+
+Execution
+---------
+
+The planned inserts are then handed off to the *executor*, which executes them
+across the cluster.
+
+This process is managed by the node responsible for handling the client
+request, but the operations needed to complete the inserts can be handed off to
+any `data node`_ in the cluster.
+
+How the executor proceeds depends on the type of inserts being executed.
+
+Executing Single Inserts
+........................
+
+The executor performs the following actions:
+
+1. Query the :ref:`cluster state <cluster_state_management>` for the the node
+   that holds the primary shard corresponding to the ``shardId`` of the
+   insert
+
+2. Check that :ref:`the required number of replicas <minimum_active_shards>`
+   are available
+
+3. Send the insert request to the primary shard
+
+4. Send an identical request request to the replicas of the primary shard
+
+   - If successful, a response is sent to the client reporting that one row has
+     been inserted
+
+   - Otherwise, an error (when using HTTP) or an exception (when using the
+     Postgres wire protocol) is sent to the client
+
+Executing Inserts With Multiple Value Expressions
+.................................................
+
+The executor makes a best effort to insert as many rows as possible. The
+executor will continue to try to insert rows, even when errors are encountered.
+
+For every insert request, the executor performs the following actions:
+
+1. Query the :ref:`cluster state <cluster_state_management>` for the the node
+   that holds the primary shard
+
+2. Check that :ref:`the required number of replicas <minimum_active_shards>`
+   are available
+
+4. Send an insert request to the primary shard
+
+5. Send an identical insert request to the replicas of the primary shard
+
+   - If successful, an insert counter is incremented
+
+6. A response is sent to the user reporting the value of the successful insert
+   counter
+
+.. NOTE::
+
+   If an insert request with multiple value expressions results in no
+   successful inserts, this is reported like any other result and does not
+   produce an an error (when using HTTP) or an exception (when using the
+   Postgres wire protocol).
+
+Executing Bulk Inserts
+......................
+
+The executor performs the same steps taken for inserts with multiple value
+expressions, as listed under the previous header, with one exception.
+
+Instead of reporting a single number corresponding to the number of successful
+inserts, the previously created bulk insert response object is populated with
+information about every insert attempted.
+
+The `bulk insert response`_ object is returned to the client, with a
+``rowcount`` for every row in the bulk insert request:
+
+- A ``rowcount`` of ``1`` indicates a successful insert
```

Comment:
don't get this, the rowCount will express the successfully inserted documents, so 0 or greater

Reply:
again, going from this:

https://crate.io/docs/crate/reference/en/latest/protocols/http.html#bulk-errors

I thought that this object is returning fine-grain results for every single row to be inserted. so you either get "1" row inserted or an error

Reply:
hm ok right somehow. I was a referring to a bit special case where one is mixing bulk request with multiple value inserts which we do support. like using following insert statement with bulk arguments:
```
insert into t1 (id, name) values (?, ?), (?, ?)
```

in this case every bulk result will return `2` as a rowcount if both inserts were successful. 
but I'm not sure if we should pointing out this case, maybe we can just rephrase it to be a bit more generic.

Reply:
if you do a bulk insert with multiple values, what happens when there's an error? how are errors distinguished from correct row count success?